course/versioning.py

```diff
@@ -577,4 +603,181 @@ def update_course(pctx):
 
 # }}}
 
+
+# {{{ git endpoint
+
+
+# {{{ wsgi wrapping
+
+# Nabbed from
+# https://github.com/Polyconseil/django-viewsgi/blob/master/viewsgi.py
+# (BSD-licensed)
+
+def call_wsgi_app(
+        application,  # type: dulwich.web.LimitedInputFilter
+        request,      # type: http.HttpRequest
+        prefix,       # type: Text
+        ):
+    # type: (...) -> http.HttpResponse
+
+    response = http.HttpResponse()
+
+    # request.environ and request.META are the same object, so changes
+    # to the headers by middlewares will be seen here.
+    environ = request.environ.copy()
+    #if len(args) > 0:
+    assert environ['PATH_INFO'].startswith(prefix)
+    environ['SCRIPT_NAME'] += prefix
+    environ['PATH_INFO'] = environ['PATH_INFO'][len(prefix):]
+
+    if six.PY2:
+        # Django converts SCRIPT_NAME and PATH_INFO to unicode in WSGIRequest.
+        environ['SCRIPT_NAME'] = environ['SCRIPT_NAME'].encode('iso-8859-1')
+        environ['PATH_INFO'] = environ['PATH_INFO'].encode('iso-8859-1')
+
+    headers_set = []   # type: List[Text]
+    headers_sent = []  # type: List[bool]
+
+    def write(data):
+        # type: (Text) -> None
+        if not headers_set:
+            raise AssertionError("write() called before start_response()")
+        if not headers_sent:
+            # Send headers before the first output.
+            for k, v in headers_set:
+                response[k] = v
+            headers_sent[:] = [True]
+        response.write(data)
+        # We could call response.flush() here, but is actually a no-op.
+
+    def start_response(status, headers, exc_info=None):
+        # Let Django handle all errors.
+        if exc_info:
+            raise exc_info[1].with_traceback(exc_info[2])
+        if headers_set:
+            raise AssertionError("start_response() called again "
+                                 "without exc_info")
+        response.status_code = int(status.split(' ', 1)[0])
+        headers_set[:] = headers
+        # Django provides no way to set the reason phrase (#12747).
+        return write
+
+    result = application(environ, start_response)
+    try:
+        for data in result:
+            if data:
+                write(data)
+        if not headers_sent:
+            write('')
+    finally:
+        if hasattr(result, 'close'):
+            result.close()
+
+    return response
+
+# }}}
+
+
+@csrf_exempt
+def git_endpoint(request, course_identifier, git_path):
+    # type: (http.HttpRequest, Text, Text) -> http.HttpResponse
+
+    auth_value = request.META.get("HTTP_AUTHORIZATION", None)
+
+    def unauthorized_access():
+        # type: () -> http.HttpResponse
+        realm = _("Relate direct git access")
```

Comment:
I'm sorry I don't understand. Can you clarify?

Reply:
Sure. My understanding of "realm" is that HTTP clients will use that as a key to look up which credentials to use. Since credentials are different by course, the realm should include some way of telling which course we're authenticating to.